spark/src/main/scala/org/apache/spark/sql/delta/commands/ReorgTableHelper.scala

```diff
@@ -145,8 +146,9 @@ trait ReorgTableHelper extends Serializable {
       fileStatuses.toList,
       ignoreCorruptFiles)
 
-    val converter =
-      new ParquetToSparkSchemaConverter(assumeBinaryIsString, assumeInt96IsTimestamp)
+    // Spark 4.0.1 changed the primary ctor signature (added a param), which breaks binary
+    // compatibility for code compiled against Spark 4.0.0. Use the stable SQLConf ctor instead.
+    val converter = new ParquetToSparkSchemaConverter(SQLConf.get)
```

Comment:
There does not seem to be a way to achieve 100% equivalency to the prior code using this constructor. But in spirit of the previous code it should be the current `SessionState.conf` passed in here. 
```suggestion
    val converter = new ParquetToSparkSchemaConverter(spark.sessionState.conf)
```

Reply:
Hey @johanl-db ,

Hey @johanl-db ! I initially  tried modifying the code same as the above:

```
val sqlConf = spark.sessionState.conf
    files.toDF(spark).as[AddFile].mapPartitions { iter =>
        filterParquetFiles(
          sqlConf,
          iter.toList,
          dataPath,
          serializedConf.value,
          ignoreCorruptFiles)(filterFileFn).toIterator
    }.collect()
```

But when we I run the testSuite i  get the below error in `TypeWideningTableFeaturePreviewSuite` and `RowTrackingPurgeSuite`
```
./build/sbt "spark/testOnly org.apache.spark.sql.delta.DeltaGenerateSymlinkManifestSuite org.apache.spark.sql.delta.typewidening.TypeWideningTableFeaturePreviewSuite org.apache.spark.sql.delta.rowid.RowTrackingPurgeSuite"
```
**_ERROR:_**
```
[info]   org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 158.0 failed 1 times, most recent failure: Lost task 0.0 in stage 158.0 (TID 145) (192.168.0.103 executor driver): java.lang.NullPointerException: Cannot invoke "org.apache.spark.internal.config.ConfigReader.get(String)" because "reader" is null
[info] 	at org.apache.spark.internal.config.ConfigEntry.readString(ConfigEntry.scala:96)
[info] 	at org.apache.spark.internal.config.ConfigEntryWithDefault.readFrom(ConfigEntry.scala:141)
[info] 	at org.apache.spark.sql.internal.SQLConf.getConf(SQLConf.scala:6661)
[info] 	at org.apache.spark.sql.internal.SQLConf.isParquetBinaryAsString(SQLConf.scala:6120)
[info] 	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.<init>(ParquetSchemaConverter.scala:64)
[info] 	at org.apache.spark.sql.delta.commands.ReorgTableHelper.filterParquetFiles(ReorgTableHelper.scala:154)
[info] 	at org.apache.spark.sql.delta.commands.ReorgTableHelper.filterParquetFiles$(ReorgTableHelper.scala:126)
[info] 	at org.apache.spark.sql.delta.commands.DeltaPurgeOperation.filterParquetFiles(DeltaReorgTableCommand.scala:108)
[info] 	at org.apache.spark.sql.delta.commands.ReorgTableHelper.$anonfun$filterParquetFilesOnExecutors$1(ReorgTableHelper.scala:122)
[info] 	at org.apache.spark.sql.execution.MapPartitionsExec.$anonfun$doExecute$3(objects.scala:198)
[info] 	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:901)
[info] 	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:901)
[info] 	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
[info] 	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
[info] 	at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
[info] 	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
[info] 	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
[info] 	at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
[info] 	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
[info] 	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
[info] 	at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
[info] 	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:93)
[info] 	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:171)
[info] 	at org.apache.spark.scheduler.Task.run(Task.scala:147)
[info] 	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$5(Executor.scala:647)
[info] 	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:80)
[info] 	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:77)
[info] 	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:99)
[info] 	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:650)
[info] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[info] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[info] 	at java.base/java.lang.Thread.run(Thread.java:840)
[info] 
[info] Driver stacktrace:
[info]   at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$3(DAGScheduler.scala:2935)
[info]   at scala.Option.getOrElse(Option.scala:201)
[info]   at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2935)
[info]   at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2927)
[info]   at scala.collection.immutable.List.foreach(List.scala:334)
[info]   at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2927)
[info]   at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1295)
[info]   at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1295)
[info]   at scala.Option.foreach(Option.scala:437)
[info]   at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1295)
[info]   at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:3207)
[info]   at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:3141)
[info]   at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:3130)
[info]   at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:50)
[info]   at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:1009)
[info]   at org.apache.spark.SparkContext.runJob(SparkContext.scala:2484)
[info]   at org.apache.spark.SparkContext.runJob(SparkContext.scala:2505)
[info]   at org.apache.spark.SparkContext.runJob(SparkContext.scala:2524)
[info]   at org.apache.spark.SparkContext.runJob(SparkContext.scala:2549)
[info]   at org.apache.spark.rdd.RDD.$anonfun$collect$1(RDD.scala:1057)
[info]   at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
[info]   at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
[info]   at org.apache.spark.rdd.RDD.withScope(RDD.scala:417)
[info]   at org.apache.spark.rdd.RDD.collect(RDD.scala:1056)
[info]   at org.apache.spark.sql.execution.SparkPlan.executeCollect(SparkPlan.scala:462)
[info]   at org.apache.spark.sql.classic.Dataset.collectFromPlan(Dataset.scala:2244)
[info]   at org.apache.spark.sql.classic.Dataset.$anonfun$collect$1(Dataset.scala:1482)
[info]   at org.apache.spark.sql.classic.Dataset.$anonfun$withAction$2(Dataset.scala:2234)
[info]   at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:654)
[info]   at org.apache.spark.sql.classic.Dataset.$anonfun$withAction$1(Dataset.scala:2232)
[info]   at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$8(SQLExecution.scala:163)
[info]   at org.apache.spark.sql.execution.SQLExecution$.withSessionTagsApplied(SQLExecution.scala:272)
[info]   at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$7(SQLExecution.scala:125)
[info]   at org.apache.spark.JobArtifactSet$.withActiveJobArtifactState(JobArtifactSet.scala:94)
[info]   at org.apache.spark.sql.artifact.ArtifactManager.$anonfun$withResources$1(ArtifactManager.scala:112)
[info]   at org.apache.spark.sql.artifact.ArtifactManager.withClassLoaderIfNeeded(ArtifactManager.scala:106)
[info]   at org.apache.spark.sql.artifact.ArtifactManager.withResources(ArtifactManager.scala:111)
[info]   at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$6(SQLExecution.scala:125)
[info]   at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:295)
[info]   at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$1(SQLExecution.scala:124)
[info]   at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:804)
[info]   at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId0(SQLExecution.scala:78)
[info]   at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:237)
[info]   at org.apache.spark.sql.classic.Dataset.withAction(Dataset.scala:2232)
[info]   at org.apache.spark.sql.classic.Dataset.collect(Dataset.scala:1482) 
[info]    at org.apache.spark.sql.delta.commands.ReorgTableHelper.filterParquetFilesOnExecutors(ReorgTableHelper.scala:116)
[info]   at org.apache.spark.sql.delta.commands.ReorgTableHelper.filterParquetFilesOnExecutors$(ReorgTableHelper.scala:103)
[info]   at org.apache.spark.sql.delta.commands.DeltaPurgeOperation.filterParquetFilesOnExecutors(DeltaReorgTableCommand.scala:108)
[info]   at org.apache.spark.sql.delta.commands.DeltaPurgeOperation.filterFilesToReorg(DeltaReorgTableCommand.scala:116)
[info]   at org.apache.spark.sql.delta.commands.OptimizeExecutor.$anonfun$optimize$1(OptimizeTableCommand.scala:307)
[info]   at org.apache.spark.sql.delta.metering.DeltaLogging.recordFrameProfile(DeltaLogging.scala:171)
[info]   at org.apache.spark.sql.delta.metering.DeltaLogging.recordFrameProfile$(DeltaLogging.scala:169)
[info]   at org.apache.spark.sql.delta.commands.OptimizeExecutor.recordFrameProfile(OptimizeTableCommand.scala:251)
[info]   at org.apache.spark.sql.delta.metering.DeltaLogging.$anonfun$recordDeltaOperationInternal$1(DeltaLogging.scala:139)
[info]   at com.databricks.spark.util.DatabricksLogging.recordOperation(DatabricksLogging.scala:128)

```

This is because when **SQLConf** instance is getting captured and serialized into the mapPartitions, it contains a config reader which doesn't get serialized and  **ConfigEntry.readString** is returning a **null**
and hence i had switched to using `SparkSession.active.sessionState.conf`